File: src/services/InputPreprocessor.js
import fs from 'fs/promises';
import path from 'path';
import chalk from 'chalk';

export class InputPreprocessor {
    static async process(input) {
        // Regex to find @filename. valid chars: alphanumeric, ., /, _, -
        const regex = /@([a-zA-Z0-9_./-]+)/g;
        const fileContents = [];

        const matches = [...input.matchAll(regex)];

        if (matches.length === 0) return input;

        for (const m of matches) {
            const token = m[0]; // @file
            const filePath = m[1]; // file

            try {
                const resolvedPath = path.resolve(process.cwd(), filePath);
                const stats = await fs.stat(resolvedPath);
                if (stats.isFile()) {
                    const content = await fs.readFile(resolvedPath, 'utf8');
                    fileContents.push(`\n--- Content from ${filePath} ---\n${content}\n--- End of ${filePath} ---\n`);
                    console.log(chalk.gray(`✔ Read file: ${filePath}`));
                } else {
                    console.warn(chalk.yellow(`⚠ Warning: ${filePath} is not a file.`));
                }
            } catch (err) {
                console.warn(chalk.yellow(`⚠ Warning: Could not read ${filePath}: ${err.message}`));
            }
        }

        if (fileContents.length > 0) {
            return input + '\n' + fileContents.join('\n');
        }

        return input;
    }
}

File: nebulus_atom/services/input_preprocessor.py
import re
from pathlib import Path


class InputPreprocessor:
    @staticmethod
    def process(user_input: str) -> str:
        regex = r"@([a-zA-Z0-9_./-]+)"
        matches = re.finditer(regex, user_input)

        file_contents = []

        for m in matches:
            file_path = m.group(1)
            try:
                resolved_path = Path.cwd() / file_path
                if resolved_path.is_file():
                    content = resolved_path.read_text(encoding="utf-8")
                    file_contents.append(
                        f"\n--- Content from {file_path} ---\n{content}\n--- End of {file_path} ---\n"
                    )
                    print(f"✔ Read file: {file_path}")
                else:
                    print(f"⚠ Warning: {file_path} is not a file.")
            except Exception as e:
                print(f"⚠ Warning: Could not read {file_path}: {e}")

        if file_contents:
            return user_input + "\n" + "".join(file_contents)

        return user_input
